### ISSUE TITLE
Developer want to rework client and server modules/classes

### ISSUE THREAD

**Issue by Ismoh** [Ismoh/NoitaMP#48]
This is the successor issue to #27, which is more or less done and would be to huge to continue. 

- [x] Rename server_class to Server
- check for inheritance of current server.self implementation to avoid a Server 'object' in Server.lua and sock.lua
- clients store last received nuid to see check that nothing is out of sync -> does this make sense? Maybe check if there is already a globals set. Mhh. Check if there is a gap > 2 between latest stored nuid and next received and log a warning then. 
- [x] NetworkVscUtils
- addNetworkVscs needs to be accessible by client AND server, but! if server then nuid ~= nil. if client then nuid can be nil, because client entities like projectiles needs to ask for a nuid on server
- [x] EntityUtils
- sync entities transform and velocity to clients when spawning
- sync transform afterwards
- sync health
- sync death

TBC

**Comment by Ismoh** [comment]
FYI: based on #27 branch

**Comment by Ismoh** [comment]
Use reliable and unreliable channels.

Reliable:
- entity spawn
- entity kill
- health
- world sync

Unreliable
- velocity
- position
-

**Comment by Ismoh** [comment]
- Send ACK per received message.
- Store nuid per peer and resend if not ACKed
- does a message needs a id?

**Comment by Ismoh** [comment]
Before doing any network stuff, extend logger with categoriea like GUID, NUID, ENTITY, NET, NOITA_COMPONENTS, etc.

Add log level per category in modSettings.

**Comment by Ismoh** [comment]
Change log of client server to use pprint.format to get a return instead of writing to io or console:
https://github.com/jagt/pprint.lua#options

**Comment by Ismoh** [comment]
Remove all callbacks and rework them step by step, also schemas.
When connecting dont send minä, sent seed and restart client, then auto connect and sent minä.

**Comment by Ismoh** [comment]
Use apto and exe filename for unique guid

**Comment by Ismoh** [comment]
Put networking update in one function and run this is WorldPreUpdate and OnPausePreUpdate.
Does it make sense to run Networking update on pre normal and postUpdate?

**Comment by Ismoh** [comment]
Add saveSlot number into saveSlot cache.

**Comment by Ismoh** [comment]
When connecting the second time, clients save directory has to be empty, otherwise seed doesnt work:
Try to remove the savefolders content with start /wait, because lfs isnt able to delete. I think because the directory is still in use. see below:
https://stackoverflow.com/a/13258451/3493998

**Comment by Ismoh** [comment]
![Uploading unknown-1.png…]()

**Comment by Ismoh** [comment]
![unknown-1](https://user-images.githubusercontent.com/12631485/170495691-13e41a0d-2ca0-4ed3-b065-8588ff018198.png)

**Comment by Ismoh** [comment]
Fix unit test! 😡

**Comment by Ismoh** [comment]
Add version to playerInfo to detect version missmatch

**Comment by Ismoh** [comment]
Add exclude include for spawning entities. Player for example do need to be spawned, but a dump copy of it.

When server got needNuid, check if entity already was spawned (client guid + localEntityId).

I think I dont need messageIds, when entityIds + guid can be used.

Add reseting guid, when development build is used, otherwise guid will be the same. Do I want to add guid check for server/client? Yes...

**Comment by Ismoh** [comment]
When sock.lua creates a new client on server object, use NoitaMP client object, instead of sock.lua client object.

**Comment by Ismoh** [comment]
Override all send functions to set acknowledge. Copy&paste acknowledge of sock.lua#Client:send

**Comment by Ismoh** [comment]
I need to add a 'cache' list for detecting, if entities moved and needs to be synced:
entities[entityId] = { health, pos, velo, etc }
Then getFilteredEntities
After that compare
In addition: use callback function in getFilteredEntities to add additionalChecks like:
- nuid set?
- moved/changed?
- isRelatedToMinä (inventory)

**Comment by Ismoh** [comment]
Only owner is allowed to send entityData?

**Comment by Ismoh** [comment]
Load entities once per world update and pass the list as parameter.?

**Comment by Ismoh** [comment]
- SendVelocity and Position when created
- SendVelocity only once? Does it change? 
- SendPosition each x frames
- Change 'DoesAnythingChanged' to be more inaccurate (0.2 instead of 0.01) otherwise it will be too much for client.
- Split Velo and Position, because Position is the truth and velo is only for movement and prediction. 
- send only position for minä (and maybe enemies)?

**Comment by Ismoh** [comment]
![Screenshot_20220806_003046_com discord](https://user-images.githubusercontent.com/12631485/183220000-e82a27ad-8b24-4a09-b316-9da63a36337a.jpg)

**Comment by Ismoh** [comment]
Added tick rate and changed detection of changes, but clients fps still drop to ~29-55.
I need to profile again!

**Comment by Ismoh** [comment]
- GetEntitiesInRadius + cache
- isAlive if not remove from cache, then next
- hasNuidComp, then hasChanged. If not then check if networkEntity, if not next else add nvscs
- hasChanged and iAmOwner then send new pos+velo
- add to cache

**Comment by Ismoh** [comment]
Use coroutine as iterator for
- init network entities
- sending network messages on specific tick rate (ms)
- receive network messages on preUpdate
- send network messages on postUpdate

**Comment by Ismoh** [comment]
- Cache all modSettings by putting those into config.lua
- put "doesModSettingsChange" into PauseCallback funtion of init.lua, so that the check isnt done each update while playing.
- doesModSettingsChange only needs to check config.lua with modSettings
- rename all ModSettings to camelCase so that I am able to add better reading mapping:
config.lua:
```lua
modSettings {
radiusToDetectEntities.key = "noita-mp.radiusToDetectEntities",
radiusToDetectEntities.values = 250,
...
}
```

**Comment by Ismoh** [comment]
First check how long it takes to get a mod setting, by profiling ModSettingGet("foo.bar"):
```lua
local cpc = CustomProfiler.start("ModSettinGet")
local foo = ModSettingGet("bar.z")
CustomProfiler.stop("ModSettingGet", cpc)
```

**Comment by Ismoh** [comment]
> First check how long it takes to get a mod setting, by profiling ModSettingGet("foo.bar"):
> 
> ```lua
> local cpc = CustomProfiler.start("ModSettinGet")
> local foo = ModSettingGet("bar.z")
> CustomProfiler.stop("ModSettingGet", cpc)
> ```

ModSettings execution is below 1ms! So this isn't the problem.

**Comment by Ismoh** [comment]
> Use coroutine as iterator for
> 
> * init network entities
> * sending network messages on specific tick rate (ms)
> * receive network messages on preUpdate
> * send network messages on postUpdate

Yield return iterator or index when execution time is greater equals 25ms (=40fps)
Next frame continue on for loop

**Comment by Ismoh** [comment]
https://love2d.org/forums/viewtopic.php?t=79980

**Comment by Ismoh** [comment]
Profiler:
Windows:
https://stackoverflow.com/questions/6207573/how-to-append-output-to-the-end-of-a-text-file

Linux:
https://stackoverflow.com/questions/17701989/how-do-i-append-text-to-a-file

**Comment by Ismoh** [comment]
Looks like I found the issue?
I'll see, if this is the problem the next days:
![image](https://user-images.githubusercontent.com/12631485/195954321-e67d269a-73a9-4727-b9bd-79ddb4ef6eac.png)
It has to be `EntityUtils.syncEntityData.for`!

If you are curious, have a look on frame 4559 in report.html: [reports.zip](https://github.com/Ismoh/NoitaMP/files/9790608/reports.zip)

**Comment by Ismoh** [comment]
Clear caches, when entering and exiting holy mountain!

**Comment by Ismoh** [comment]
Exclude minäs when syncing entities. Think of client is in servers entity sync radius, this client dont want to get servers position of his own copy on the server!

**Comment by Ismoh** [comment]
Receive network messages in preworldupdate and send messages in postworldupdate?

**Comment by Ismoh** [comment]
![image](https://user-images.githubusercontent.com/12631485/200199919-13e7b5e7-283c-459b-81d2-dfcf10e4dfaf.png)

**Comment by Ismoh** [comment]
Note:
If entity has children, loop children and add nuid + send it two clients

**Comment by ofoxsmith** [comment]
Shouldn't this branch be merged into develop? The goals of this PR have been completed, so why hasn't it been merged?

**Comment by ofoxsmith** [comment]
As far as I can tell, most of the issues and PRs based off of this branch are seperate issues - not related to the client/server module rework.

**Comment by Ismoh** [comment]
It isn't merged, because other branches are based on this. But yes, it's time for it

**Comment by ofoxsmith** [comment]
Should we merge it after #89 is finished?
